proposals/secure-storage.md

```diff
@@ -0,0 +1,158 @@
+
+```
+{
+  type: "MACOS_KEYCHAIN",
+  availableAuthentication: [
+    "PIN",
+    "PASSWORD",
+    "BIOMETRY_FACE",
+    "BIOMETRY_FINGERPRINT"
+  ]
+}
+```
+
+**browser.secureStorage.generateKey**
+
+The generateKey API allows you to generate a key to be used for protecting your data, choosing which authentication types you will allow for retrieving it in the future.
+
+
+```
+const key = browser.secureStorage.generateKey({
+  id: "example-key"
+  type: "RSA",
+  authentication: ["BIOMETRY_FACE", "BIOMETRY_FINGERPRINT"]
+});
+```
+
+**browser.secureStorage.retrieveKey**
+
+Finally, you can retrieve the key. The browser will only provide it if the user authenticates with one of the allowed mechanisms for this secret, and will throw an error otherwise.
+
+```
+const key = browser.secureStorage.retrieveKey({ id: "example-key" });
+```
+
+### Proposal 2
+
+This proposal allows a string to be stored. It is much simpler for developers to consume, but adds complexity for browser vendors, as it is further from the hardware. On platforms other than macOS the browser must create a key behind the scenes to encrypt the data and then store it on disk. It must remember which key the data was encrypted with and use this to decrypt it in the future.
+
+**browser.secureStorage.getInfo**
+
+This is implemented identically to Proposal 1.
+
+**browser.secureStorage.store**
+
+This stores the provided string.
+
+```
+browser.secureStorage.store({
+  id: "example-data"
+  authentication: ["BIOMETRY_FACE", "BIOMETRY_FINGERPRINT"],
+  data: JSON.stringify({ password: "!72AH8d_.-*gFgNFPUFz2" })
+});
+```
+
+**browser.secureStorage.retrieve**
+
+This retrieves the stored data. The browser will only provide it if the user authenticates with one of the allowed mechanisms for this secret, and will throw an error otherwise.
+
+```
+browser.secureStorage.retrieve({ id: "example-data" });
+```
+
+## FAQ
+
+**How large could the stored data be?**
+
+The expectation is that implementation specific limits would be set. Only a very small allowance (< 1KB) is needed for the known use cases.
+
+**Are there any extensions that would use this?**
+
+This proposal is drafted by the team at 1Password. Our [extension](https://chrome.google.com/webstore/detail/1password-%E2%80%93-password-mana/aeblfdkhhhdcdjpifhhbdiojplfjncoa?hl=en) is available for Chrome, Firefox, Safari and Edge and has more than 2M weekly active users in the Chrome Web Store alone. We would be quick to implement this in conjunction with or in place of our current Native Messaging solution.
+
+The Dashlane browser extension, also a password manager with over 4M weekly active users in the Chrome Web Store, currently provides a remember me and biometric authentication feature in its extension to make repeated authentication easier. They would also be very quick to implement this in place of their solution which requires a network connection, as they don’t leverage a native application to interface with native storage.
```

Comment:
Do we have a public comment of someone from Dashlane supporting this proposal?  If not, perhaps better to leave this for a later PR they could submit.

Reply:
@bradcrush reached out out-of-band and shared this quote which he was happy for us to use in the proposal. I'm fairly sure he's happy with it being included here but I'll ask if he can drop an approval so we have everything documented in one place.

Reply:
This quote is good for us, confirming here publicly on behalf of Dashlane, thanks!